File: apps/vnext-web/lib/vnext/journal.ts
import type { EngineEvent, Item, Session, Turn, TurnState, UsageSnapshot } from "@telar/engine-client";

/**
 * The client-side fold over protocol v2's journal.
 *
 * WHAT CHANGED FROM v1: the fold used to produce `{ prompt, text }` — one
 * string of accumulated prose per turn — because that is all the engine sent.
 * v2 carries a timeline, so a turn now owns an ordered list of ITEMS (tool
 * calls, reasoning, assistant messages) and streaming text lands on the item it
 * belongs to instead of being concatenated into one blob.
 *
 * THE FOLD IS THE ONLY SOURCE OF TRUTH FOR ORDER. Items sort by the event id
 * that opened them, never by timestamp: two events can share a millisecond, and
 * `at` is when the engine recorded a row rather than when the provider produced
 * it. The id is monotonic per session by construction.
 */

export type JournalItem = Item & {
  /** Deltas accumulated in arrival order. Empty for items that never stream. */
  streamedText: string;
  /** The event id that opened this item — the sort key, not a display value. */
  openedBy: number;
};

export type JournalTurn = {
  runId: string;
  prompt: string;
  state: TurnState;
  items: JournalItem[];
  /** The assistant's final text, as the engine recorded it on completion. */
  resultText: string;
  failure?: string;
  usage?: UsageSnapshot;
};

/** Merges a cursor page without duplicating durable journal records. */
export function appendJournalEvents(existing: EngineEvent[], incoming: EngineEvent[]): EngineEvent[] {
  const events = new Map(existing.map((event) => [event.id, event]));
  for (const event of incoming) events.set(event.id, event);
  return [...events.values()].sort((left, right) => left.id - right.id);
}

export function journalCursor(events: EngineEvent[]): number {
  return events.reduce((cursor, event) => Math.max(cursor, event.id), 0);
}

/**
 * Fold turns, the engine's item projection, and the event tail into a
 * renderable transcript.
 *
 * ITEMS COME FROM BOTH SIDES ON PURPOSE. The snapshot's `items` is what makes
 * opening a long session cheap — no replay from event 1 — while the event tail
 * is what makes a live turn stream. Applying the tail second means a row the
 * snapshot caught mid-flight is corrected by the events that followed it.
 */
export function projectJournal(turns: Turn[], items: Item[], events: EngineEvent[]): JournalTurn[] {
  const byRun = new Map<string, JournalTurn>(
    turns.map((turn) => [
      turn.runId,
      {
        runId: turn.runId,
        prompt: turn.input,
        state: turn.state,
        items: [],
        resultText: turn.resultText ?? "",
        ...(turn.failure ? { failure: turn.failure.message } : {}),
        ...(turn.usage ? { usage: turn.usage } : {}),
      },
    ]),
  );
  const seenItems = new Map<string, JournalItem>();

  const upsert = (item: Item, openedBy: number): JournalItem | undefined => {
    const turn = byRun.get(item.runId);
    if (!turn) return undefined;
    const existing = seenItems.get(item.id);
    const merged: JournalItem = {
      ...item,
      streamedText: existing?.streamedText ?? "",
      openedBy: existing?.openedBy ?? openedBy,
    };
    seenItems.set(item.id, merged);
    const index = turn.items.findIndex((candidate) => candidate.id === item.id);
    if (index === -1) turn.items.push(merged);
    else turn.items[index] = merged;
    return merged;
  };

  // The snapshot first, so a session that has been running for an hour renders
  // without replaying its whole journal. Event id 0 is not a real id, so these
  // sort before anything the tail opens — which is the correct relative order
  // for rows that already existed when the page loaded.
  for (const item of items) upsert(item, 0);

  for (const event of events) {
    const turn = event.runId ? byRun.get(event.runId) : undefined;

    switch (event.type) {
      case "turn.accepted":
        if (!byRun.has(event.turn.runId)) {
          byRun.set(event.turn.runId, {
            runId: event.turn.runId,
            prompt: event.turn.input,
            state: event.turn.state,
            items: [],
            resultText: "",
          });
        }
        break;
      case "turn.claimed":
        if (turn) turn.state = "claimed";
        break;
      case "turn.started":
        if (turn) turn.state = "running";
        break;
      case "turn.requeued":
        if (turn) turn.state = "queued";
        break;
      case "turn.completed":
        if (turn) {
          turn.state = "completed";
          turn.resultText = event.resultText;
          if (event.usage) turn.usage = event.usage;
        }
        break;
      case "turn.failed":
        if (turn) {
          turn.state = "failed";
          turn.failure = event.message;
        }
        break;
      case "turn.stopped":
        if (turn) turn.state = "stopped";
        break;
      case "turn.ambiguous":
        if (turn) turn.state = "ambiguous";
        break;
      case "turn.discarded":
        if (turn) turn.state = "discarded";
        break;
      case "item.started":
      case "item.updated":
      case "item.completed":
        upsert(event.item, event.id);
        break;
      case "content.delta": {
        // A delta for an item this client has not seen is DROPPED, not
        // buffered. It means the fold is missing the row that opened it, and
        // inventing a placeholder would render a message with no idea what
        // kind of row it belongs to. The next snapshot repairs it.
        const item = seenItems.get(event.itemId);
        if (item) item.streamedText += event.text;
        break;
      }
      case "usage.updated":
        if (turn) turn.usage = event.usage;
        break;
      default:
        // Every other family (runtime.*, request.*, task.*, browser.*, mcp.*)
        // is contract but not yet rendered. Ignoring them here is deliberate;
        // dropping them at PARSE time would not be, which is why
        // safeParseEvent only skips rows it cannot understand at all.
        break;
    }
  }

  for (const turn of byRun.values()) {
    turn.items.sort((left, right) => left.openedBy - right.openedBy || left.startedAt - right.startedAt);
  }
  return [...byRun.values()];
}

export function isActiveTurn(state: TurnState): boolean {
  return state === "queued" || state === "claimed" || state === "running";
}

/**
 * The text a rendered item should show.
 *
 * Streamed deltas WIN over the stored detail while a turn is live, because the
 * engine only folds accumulated text into `items.json` when the item closes —
 * writing the whole document per token would be absurd. Once closed, the two
 * agree and either is correct.
 */
export function itemText(item: JournalItem): string {
  if (item.streamedText) return item.streamedText;
  if (item.detail.type === "assistant_message" || item.detail.type === "reasoning") return item.detail.text;
  if (item.detail.type === "user_message") return item.detail.text;
  return "";
}

/** A one-line label for a collapsed row, preferring what the engine stored. */
export function itemLabel(item: JournalItem): string {
  if (item.title) return item.title;
  switch (item.detail.type) {
    case "command_execution":
      return item.detail.command.command || "command";
    case "file_change":
      return item.detail.change.path;
    case "file_read":
      return item.detail.read.path;
    case "mcp_tool_call":
    case "dynamic_tool_call":
    case "browser_action":
      return item.detail.call.name;
    case "web_search":
      return item.detail.query;
    case "error":
      return item.detail.error.message;
    default:
      return item.detail.type;
  }
}

/** Rows that render as a tool card rather than as prose. */
export function isToolItem(item: JournalItem): boolean {
  return (
    item.detail.type === "command_execution" ||
    item.detail.type === "file_change" ||
    item.detail.type === "file_read" ||
    item.detail.type === "mcp_tool_call" ||
    item.detail.type === "dynamic_tool_call" ||
    item.detail.type === "web_search" ||
    item.detail.type === "browser_action"
  );
}

/** The output body of a finished tool call, when it has one. */
export function toolOutput(item: JournalItem): string | undefined {
  if (item.detail.type === "command_execution") return item.detail.command.outputPreview;
  if (item.detail.type === "mcp_tool_call" || item.detail.type === "dynamic_tool_call") {
    const output = item.detail.call.output;
    return typeof output === "string" ? output : output === undefined ? undefined : JSON.stringify(output, null, 2);
  }
  return undefined;
}

export type { Session };

File: apps/ios/TelarMobile/Sync/Journal.swift
import Foundation

/// Port of `apps/web/lib/engine/journal.ts` — the client-side fold over the
/// session's event journal. Like the web, the fold is RECOMPUTED from
/// (turns, items, events, tasks) rather than applied incrementally: the event
/// list is bounded per session, the recompute is the oracle the web tests pin,
/// and an incremental fold is a second implementation that can drift.
///
/// The TS fold leans on JS reference semantics — the object in `seenItems`
/// IS the object in the rendered list, so a delta append lands in both. The
/// Swift port keeps that with internal reference boxes and materialises value
/// types at the end.

struct JournalItem: Identifiable, Equatable {
    var item: Item
    /// Deltas accumulated in arrival order. Empty for items that never stream.
    var streamedText: String
    /// The event id that opened this item — the sort key, never timestamps:
    /// two events can share a millisecond; the id is monotonic by construction.
    var openedBy: Int

    var id: EngineID { item.id }
    var status: ItemStatus { item.status }
    var detail: ItemDetail { item.detail }

    /// Streamed deltas win over the stored detail while live — the engine only
    /// folds text into the item when it closes. Once closed the two agree.
    var text: String {
        if !streamedText.isEmpty { return streamedText }
        switch item.detail {
        case .assistantMessage(let text), .reasoning(let text), .userMessage(let text):
            return text
        default:
            return ""
        }
    }

    /// One-line label for a collapsed row, preferring what the engine stored —
    /// the engine owns collapsed-row labels; clients must not invent them.
    var label: String {
        if let title = item.title, !title.isEmpty { return title }
        switch item.detail {
        case .commandExecution(let command): return command.command.isEmpty ? "command" : command.command
        case .fileChange(let change): return change.path
        case .fileRead(let read): return read.path
        case .mcpToolCall(let call), .dynamicToolCall(let call): return displayToolName(call.name)
        case .browserAction(let call, _): return displayToolName(call.name)
        case .webSearch(let query, _): return query
        case .error(let error): return error.message
        case .unknown(let label): return label ?? "unknown"
        case .userMessage: return "user_message"
        case .assistantMessage: return "assistant_message"
        case .reasoning: return "reasoning"
        case .plan: return "plan"
        case .task: return "task"
        case .contextCompaction: return "context_compaction"
        }
    }

    /// The output body of a finished tool call, when it has one.
    var toolOutput: String? {
        switch item.detail {
        case .commandExecution(let command):
            return command.outputPreview
        case .mcpToolCall(let call), .dynamicToolCall(let call), .browserAction(let call, _):
            switch call.output {
            case .string(let text): return text
            case .none: return nil
            case .some(let value): return value.prettyPrinted
            }
        default:
            return nil
        }
    }
}

struct JournalTask: Identifiable, Equatable {
    var task: AgentTask
    var items: [JournalItem]
    var id: EngineID { task.id }
}

struct JournalTurn: Identifiable, Equatable {
    var runId: EngineID
    var prompt: String
    /// The compaction gesture — a system row, not a bubble.
    var isCompactGesture: Bool = false
    var state: TurnState
    /// The MAIN LOOP's timeline only — sub-agent rows live on `tasks`.
    var items: [JournalItem]
    var tasks: [JournalTask]
    var startedAt: Timestamp?
    /// When anything last happened — deltas included. What "gone quiet" is
    /// measured from; only a gap since the LAST event can tell slow from stuck.
    var lastActivityAt: Timestamp?
    var resultText: String
    var failure: String?
    var usage: UsageSnapshot?

    var id: EngineID { runId }

    /// An open `context_compaction` item — the provider squeezing right now.
    var isCompacting: Bool {
        items.contains { item in
            if case .contextCompaction = item.detail { return item.status == .inProgress }
            return false
        }
    }
}

/// Strips `mcp__server__` framing — addressing, not meaning. The rest-join
/// matters: `mcp__github__fetch__pr` names a tool called `fetch__pr`.
func displayToolName(_ name: String) -> String {
    guard name.hasPrefix("mcp__") else { return name }
    let parts = name.split(separator: "__", omittingEmptySubsequences: false).map(String.init)
    guard parts.count >= 3, !parts[1].isEmpty else { return name }
    return parts.dropFirst(2).joined(separator: "__")
}

/// Merges a cursor page without duplicating durable journal records.
func appendJournalEvents(_ existing: [EngineEvent], _ incoming: [EngineEvent]) -> [EngineEvent] {
    var byId = [Int: EngineEvent]()
    for event in existing { byId[event.id] = event }
    for event in incoming { byId[event.id] = event }
    return byId.values.sorted { $0.id < $1.id }
}

func journalCursor(_ events: [EngineEvent]) -> Int {
    events.reduce(0) { max($0, $1.id) }
}

// MARK: - the fold

private final class ItemBox {
    var item: Item
    var streamedText = ""
    var openedBy: Int
    init(item: Item, openedBy: Int) {
        self.item = item
        self.openedBy = openedBy
    }
}

private final class TaskBox {
    var task: AgentTask
    var items: [ItemBox] = []
    init(task: AgentTask) { self.task = task }
}

private final class TurnBox {
    var runId: EngineID
    var prompt: String
    var isCompactGesture: Bool
    var state: TurnState
    var items: [ItemBox] = []
    var tasks: [TaskBox] = []
    var startedAt: Timestamp?
    var lastActivityAt: Timestamp?
    var resultText: String
    var failure: String?
    var usage: UsageSnapshot?
    init(turn: Turn) {
        runId = turn.runId
        prompt = turn.input
        isCompactGesture = turn.kind == "compact"
        state = turn.state
        startedAt = turn.startedAt
        resultText = turn.resultText ?? ""
        failure = turn.failure?.message
        usage = turn.usage
    }
}

func projectJournal(
    turns: [Turn], items: [Item], events: [EngineEvent], tasks: [AgentTask] = []
) -> [JournalTurn] {
    var runOrder: [EngineID] = []
    var byRun = [EngineID: TurnBox]()
    for turn in turns where byRun[turn.runId] == nil {
        byRun[turn.runId] = TurnBox(turn: turn)
        runOrder.append(turn.runId)
    }
    var seenItems = [EngineID: ItemBox]()
    var seenTasks = [EngineID: TaskBox]()

    func upsertTask(_ task: AgentTask) {
        guard let turn = byRun[task.runId] else { return }
        if let existing = seenTasks[task.id] {
            // Items already collected survive: every task event repeats the
            // whole task, and a replace would empty the list each time.
            existing.task = task
        } else {
            let box = TaskBox(task: task)
            seenTasks[task.id] = box
            turn.tasks.append(box)
        }
    }

    func upsert(_ item: Item, openedBy: Int) {
        guard let turn = byRun[item.runId] else { return }
        let owner = item.taskId.flatMap { seenTasks[$0] }
        if let existing = seenItems[item.id] {
            existing.item = item
            // openedBy and streamedText survive item.updated/completed.
            // If the row was parked on the main timeline because its task had
            // not been met yet, move it home now (the web keeps a stale copy
            // and lets the next snapshot repair it; a move is the same repair
            // without the duplicate row).
            if let owner, !owner.items.contains(where: { $0 === existing }) {
                turn.items.removeAll { $0 === existing }
                owner.items.append(existing)
            }
        } else {
            let box = ItemBox(item: item, openedBy: openedBy)
            seenItems[item.id] = box
            // A row filed under a task the fold has not met stays on the MAIN
            // timeline rather than being dropped — an invisible row is worse
            // than a misplaced one.
            if let owner { owner.items.append(box) } else { turn.items.append(box) }
        }
    }

    // The snapshot first: opening a long session must not replay its journal.
    // openedBy 0 is not a real id, so snapshot rows sort before anything the
    // tail opens. Tasks BEFORE items, so sub-agent rows find their owner.
    for task in tasks { upsertTask(task) }
    for item in items { upsert(item, openedBy: 0) }
    // The quiet clock, seeded so a page opened onto a running turn does not
    // start by claiming it has been silent since it began.
    for runId in runOrder {
        guard let turn = byRun[runId] else { continue }
        var latest = turn.startedAt ?? 0
        for box in turn.items { latest = max(latest, box.item.completedAt ?? box.item.startedAt) }
        for box in turn.tasks { latest = max(latest, box.task.updatedAt) }
        if latest > 0 { turn.lastActivityAt = latest }
    }

    for event in events {
        let turn = event.runId.flatMap { byRun[$0] }
        // ANY event on the turn is activity, deltas included — item timestamps
        // do not move while text streams.
        if let turn { turn.lastActivityAt = max(turn.lastActivityAt ?? 0, event.at) }

        switch event.payload {
        case .turnAccepted(let accepted, _):
            if byRun[accepted.runId] == nil {
                byRun[accepted.runId] = TurnBox(turn: accepted)
                runOrder.append(accepted.runId)
            }
        case .turnCompleted(let resultText, let usage):
            guard let turn else { break }
            turn.state = .completed
            turn.resultText = resultText
            if let usage { turn.usage = usage }
        case .turnFailed(_, let message):
            guard let turn else { break }
            turn.state = .failed
            turn.failure = message
        case .turnStopped:
            turn?.state = .stopped
        case .itemStarted(let item), .itemUpdated(let item), .itemCompleted(let item):
            upsert(item, openedBy: event.id)
        case .turnPlanUpdated(let item):
            upsert(item, openedBy: event.id)
        case .contentDelta(let itemId, _, let text):
            // A delta for an unseen item is DROPPED, not buffered: the fold is
            // missing the row that opened it, and a placeholder would render a
            // message with no idea what kind of row it belongs to.
            seenItems[itemId]?.streamedText += text
        case .taskStarted(let task), .taskProgress(let task), .taskCompleted(let task):
            upsertTask(task)
        case .usageUpdated(let usage):
            turn?.usage = usage
        case .browserControlChanged(let controller):
            // The §6 marker row, mirrored from the web fold: a takeover lands
            // inside the turn it interrupted as a one-line unknown-detail row.
            // Between turns (no runId) the live badge is the story, not history.
            guard let runId = event.runId, controller != "idle" else { break }
            upsert(
                Item(
                    id: "control_\(event.id)",
                    runId: runId,
                    sessionId: event.sessionId,
                    status: .completed,
                    title: nil,
                    detail: .unknown(label: controller == "human" ? "You took the browser" : "The browser was handed back to the agent"),
                    startedAt: event.at,
                    completedAt: event.at,
                    taskId: nil
                ),
                openedBy: event.id
            )
        case .requestOpened, .requestResolved, .sessionUpdated:
            break
        case .none:
            // State transitions the payload enum does not carry ride the type
            // string — same outcomes as the web fold's switch arms.
            guard let turn else { break }
            switch event.type {
            case "turn.claimed": turn.state = .claimed
            case "turn.started":
                turn.state = .running
                turn.startedAt = turn.startedAt ?? event.at
            case "turn.requeued": turn.state = .queued
            case "turn.steering": turn.state = .steering
            case "turn.steered": turn.state = .steered
            case "turn.ambiguous": turn.state = .ambiguous
            case "turn.discarded": turn.state = .discarded
            default: break
            }
        }
    }

    func materialize(_ box: ItemBox) -> JournalItem {
        JournalItem(item: box.item, streamedText: box.streamedText, openedBy: box.openedBy)
    }
    let byOpen: (ItemBox, ItemBox) -> Bool = {
        ($0.openedBy, $0.item.startedAt) < ($1.openedBy, $1.item.startedAt)
    }
    return runOrder.compactMap { runId in
        guard let turn = byRun[runId] else { return nil }
        let sortedTasks = turn.tasks.sorted {
            ($0.task.startedAt, $0.task.id) < ($1.task.startedAt, $1.task.id)
        }
        return JournalTurn(
            runId: turn.runId,
            prompt: turn.prompt,
            isCompactGesture: turn.isCompactGesture,
            state: turn.state,
            items: turn.items.sorted(by: byOpen).map(materialize),
            tasks: sortedTasks.map { JournalTask(task: $0.task, items: $0.items.sorted(by: byOpen).map(materialize)) },
            startedAt: turn.startedAt,
            lastActivityAt: turn.lastActivityAt,
            resultText: turn.resultText,
            failure: turn.failure,
            usage: turn.usage
        )
    }
}
